### ISSUE TITLE
Apiserver to kubelet communication should not use NodeName as address.

### ISSUE THREAD

**Issue by mkulke** [kubernetes/kubernetes#22063]
In `pkg/registry/pod/strategy.go` the host address for kubelets seem to be derived from `pod.Spec.nodeName`. It should however use the `Node.Status.Addresses` field to find an IP (the docs seem to indicate `InternalIP` should be used: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/node.md#node-addresses

When a `NodeName` is resolved to an address/IP which is not open to the apiserver for various reasons (in our case `eth0` is blocked for all ingress traffic by security groups) several operations fail. This affects apiserver -> kubelet communication like exec, attach, port-forward.

I'll try addressing this with a PR, when it's acknowledged that this is a valid problem.

**Comment by ncdc** [comment]
cc @liggitt @kubernetes/rh-cluster-infra @thockin @bgrant0607 @kubernetes/rh-networking

**Comment by ncdc** [comment]
This affects at least exec, attach, and port forward. Maybe logs too.

**Comment by mkulke** [comment]
i digged a bit, this issue popped up a few times already:

#18525

The reason for this behaviour seems to be pragmatic solution for TLS issue (b/c often IPs aren't included in certs). However cloud-providers (like AWS, OpenStack) do not necessarily set resolvable NodeNames.

**Comment by bgrant0607** [comment]
This has come up many times (see #2462). 

Yes, we shouldn't try to resolve nodeName. Addresses are intended to be used. We've talked about adding more information to that array, to better indicate which address should be used by apiserver.

**Comment by Calpicow** [comment]
+1 #22109

**Comment by liggitt** [comment]
Is there a way to indicate a resolveable DNS name, for TLS purposes?

**Comment by mkulke** [comment]
Summary of what I gathered so far:

- The kubelet server is addressed by `pod.Spec.NodeName` property in `pkg/registry/pod` for apiserver => kubelet operations.
- Via `--hostname-override` you can specify the node name, otherwise the hostname is used.
- Cloud providers may populate `node.Status.Addresses` with a Node's available IP addresses. 
- API spec currently allows those of type: `Hostname, ExternalIP or InternalIP`.
- Type `Hostname` is unused (deprecated) says https://github.com/kubernetes/kubernetes/blob/master/docs/admin/node.md#node-addresses
- The kubelet server is protected by a self-signed certificate which it generates on startup, unless a cert/key pair is specified as option.
-  `util.GenerateSelfSignedCert` has a param `alternateIPs` which _could_ be used with `Node.Status.Addresses`.
- Then `pkg/registry/pod` could use `node.Status.Addresses`  via  `pod.Spec.NodeName`, instead of `NodeName` directly.
- What would be the criteria for selecting an Address? `InternalIP`?
- The self signed cert is currently generated at the very startup of the kubelet - there are no node addresses yet.

How to cope with this?

- Bring-your-own-certs when you use `--override-hostname`.
- Continue to use `NodeName` as address and `--override-hostname`. (Cloud providers sometimes use the `NodeName`as heuristic to find the nodes in the respective IaaS APIs. This may fail if you use an IP from a dedicated network, which other than the public or private ips is not a unique identifier, like `192.168.123.3`. This could be fixed in cloud-provider code I think.)

**Comment by Calpicow** [comment]
As far as AWS goes, `--hostname-override` only overrides the `kubernetes.io/hostname` node label. `NodeName` remains unchanged.

**Comment by mkulke** [comment]
@Calpicow: good point. The node name is actually gathered from the cloud-provider:

https://github.com/kubernetes/kubernetes/blob/f0c0af256110bc75447ce8997522ca1a08f58e20/cmd/kubelet/app/server.go#L578

however the cert is using `s.HostnameOverride`:

https://github.com/kubernetes/kubernetes/blob/f0c0af256110bc75447ce8997522ca1a08f58e20/cmd/kubelet/app/server.go#L379

**Comment by Calpicow** [comment]
That looks good to me. I would think the easiest fix would be to replace all `nodeHost` references to the host (or host override value):

https://github.com/kubernetes/kubernetes/blob/f0c0af256110bc75447ce8997522ca1a08f58e20/pkg/registry/pod/strategy.go#L277

**Comment by mkulke** [comment]
you mean replace them the `kubernetes.io/hostname` label?

**Comment by Calpicow** [comment]
Yes, since the certs will match up. `--hostname-override` used to alter `NodeName`, but now this logic needs to be updated since it's no longer the case.

**Comment by mkulke** [comment]
sounds like a workable solution (still need to tackle the problem with cloud-providers having discovery issues due to `--hostname-override`, but i'm pretty certain now the problem is in their domain), maybe a bit "untyped". let's gather some more input... (also: what are NodeAddresses for then?)

**Comment by liggitt** [comment]
Not sure I'm comfortable making an annotation the main way to determine the node's hostname. Seems like it should be more explicit, like bringing back the Hostname status address.

**Comment by liggitt** [comment]
Also, do we envision targeting 1.2 or after with a change here?

**Comment by Calpicow** [comment]
+1 for 1.2. This is a hard breaking change for us. It is unusable in our environment until it is fixed.

**Comment by liggitt** [comment]
> This is a hard breaking change for us. 

Is the behavior different than 1.1? Not disagreeing something is needed, just trying to determine if the issue is a regression from the previous release.

**Comment by Calpicow** [comment]
My mistake--I've reviewed the relevant parts from v1.1 and v1.2 and their behavior is similar.

I also forgot about the modified AWS cloudprovider that we use, which uses `PrivateIpAddress` as `NodeName` instead of `PrivateDnsName`. Good news is that I could sustain with this hack for now.

A brief review of the code and some PRs/issues shows that a fix isn't as trivial as I thought, and has potential to break a lot more stuff.

Related:
https://github.com/kubernetes/kubernetes/pull/17643#issuecomment-176780289
https://github.com/kubernetes/kubernetes/issues/21890

**Comment by mkulke** [comment]
I looked a bit more into the problem, it's definitely not a low hanging fruit. Changing this behavior (from using `Pod.Spec.NodeName` to `Node.Status.Addresses` of the respective node) will require changes in a lot of places. It would mean that we have to defer the TLS cert generation in kubelet until the Addresses list is populated (so we can use those as alt subject in cert). 

This list in turn is populated by cloud-providers, which somewhat arbitrarily put addresses there, not necessarily the one to which we bound the kubelet server.

I'd suggest to use the kubelet's `--address` parameter in the self-signed cert generation as an alternative subject ( the params we have available at this point). Then the kubelet logic for populating `Node.Status.Addresses` needs to be extended to add this address with a special type `KubeletIP`. The apiserver would then connect to the kubelet using this field and `Pod.Spec.NodeName` as fallback.

Makes some sense? I think it does not break anything. OTOH it's pretty immersive for something which you could consider as a merely cosmetic issue (nodename == kubelet's bound ip) or an issue in the realm of the cloud providers (should not rely on nodename for instance discovery).

**Comment by alexkappa** [comment]
Hi everyone, is there a way we can workaround this issue until this is developed? I'm using the `--cloud-provider=aws`, `--register-node=true` and `--register-schedulable=false` options on the master kubelet. No `--hostname-override` due to `--cloud-provider=aws` (#22984).

**Comment by marun** [comment]
I'm working on getting openshift running on openshift such that nodes for the nested cluster are managed via a replica set.  Since the apiserver assumes the node names are resolveable for kubelet communication, I'm having to watch for node changes and update /etc/hosts on the master node.  It would be preferable to have the apiserver use node.status.addresses since I'm ensuring that an InternalIP is being set on node registration.

**Comment by liggitt** [comment]
Using IP addresses even when a resolvable hostname is available will make TLS to nodes more difficult

**Comment by mkulke** [comment]
I also looked into populating the existing but currently unused address type `Hostname` in the `Node.Status.Addresses` array (which would respect `--hostname-override`, the kubelet's cert creation process does this already). Then `Node.Status.Addresses.Hostname` could be resolved via a Pod's `Spec.NodeName` in a log/port-forward/... API call.

However, i've seen that the Apiserver code to access objects from etcd in `pkg/registry/*` is organized very hierarchical and maps to the API calls. As far as I can tell it would not be easy to shoehorn code to retrieve a Node object in the context of a `/pod/{name}/log` call (considering authorization for objects and all, plz correct me, if I'm wrong).

So using `Pod.Spec.NodeName` to construct an URL does make sense from this perspective, but the problem remains: users need to be able to set a Node's name arbitrarily via `--hostname-override`. Currently cloudproviders (AWS, OpenStack I think) ignore it, query the IaaS API for an instance name and set the NodeName via `Instances.CurrentNodeName`.

**Comment by mkulke** [comment]
After some insights I received from @justinsb, i decided to take a stab anyway (PR #25532); i.e. removing the restriction that node names have to resolvable on the master. The changes required to achieve that are actually not very invasive.

**Comment by liggitt** [comment]
Yeah, we were already retrieving the node to look up the kubelet port, looking up a hostname override as well makes sense to me.

**Comment by anguslees** [comment]
Just fyi, there is no DNS entry for hosts in Rackspace public cloud (unless you install and run your own DNS server, edit all the /etc/hosts, etc) - and it's an optional feature for OpenStack clouds generally.

We did a bunch of work a year or two ago to introduce and use `Addresses` instead of assuming the `NodeName` was resolvable, for just this sort of case.  I'm disappointed we're here again :(
(The intended semantics of `InternalIP` vs `ExternalIP` were never defined afaik, but `Hostname` was the old/legacy behaviour of assuming you could do a naive `http://$Hostname/foo` and reach the kubelet.  This naive URL interpolation also prevented possible DNS workarounds like using a node's IPv6 literal address as the `NodeName`)

TL;DR: Please ensure that whatever we end up with this time around does not require a DNS entry for nodes.

**Comment by liggitt** [comment]
> We did a bunch of work a year or two ago to introduce and use Addresses instead of assuming the NodeName was resolvable, for just this sort of case.

How do you handle TLS with IP addresses? Do you assume nodes are serving using certs signed by a private CA that is willing to sign for IP addresses?

**Comment by mkulke** [comment]
@anguslees yup, i'm aware (been using k8s on rackspace & openstack for a while). Rackspace setups require `--hostname-override` because of this: The nodename is simply the hostname, and the hostname is a non-resolvable, dns-ified version of the instance name (e.g. my-node_1 => my-node-1). That's why currently the Rackspace cloud provider cannot set the Nodename, like other providers do.

The intention is to decouple the master => node discovery from (resolvable) Nodenames by leveraging `Node.Status.Addresses.Hostname` (which would respect `--hostname-override`).

@liggitt Currently the kubelet defaults to self-signed certificates which can have IPs as `subjectAltName`.

**Comment by liggitt** [comment]
Defaults, sure, but I wouldn't expect that to be used that way in production

**Comment by anguslees** [comment]
@liggitt: this was before TLS support came in (iirc), so that wasn't a problem ;)

So in my head we *should* have something like (correct me if I've missed anything):
- an arbitrary unique string used to identify the node internal to the kubernetes code/API
- a list of ("internal"?) addresses where the control plane can reach that node (found via `cloudprovider.NodeAddresses(name)`)
- the node has a hostname (that may be unrelated to any other id and not unique) that we don't care about
- the cloud provider can also expose its own `ExternalID` and `InstanceID` for the node (which afaics aren't used anywhere except to report (for debugging?) in API)
- we have a primary key that we associate with the node (for TLS)

Did I miss any "node IDs"?

Currently in some places in the code there is the assumption that nodename == hostname == something_that_DNS_can_turn_into_addresses, which isn't true for all providers.  It seems like we all agree we should separate that out (again) into something like the bullet points above.  The only open question remains how we get the TLS library to associate the correct primary key / CA sig with the node - is that correct?

**Comment by liggitt** [comment]
I'm referring to TLS serving by the kubelet, which has existed since 1.0. The master needs to contact the node on a hostname or IP that is valid for the serving cert presented by the node. Are you assuming the thing generating serving certs for nodes will be happy signing a cert for an IP?

**Comment by liggitt** [comment]
Also I'm not sure what you mean by "primary key associated with the node for TLS"

**Comment by anguslees** [comment]
@liggitt: urgh, sorry - corrected that to "private key"

**Comment by anguslees** [comment]
@liggitt: I don't think I understand your concern, so I'm sorry for appearing dense.  We just need something that the CA and master agree they're going to provide and look for in subject/subjectaltname - correct?  So that could be a real DNS entry (if such a thing exists), or some sort of fake hostname (which might be DNSified versions of nodename, or provider ExternalID), or IP addresses, the InstanceID "URI", or some other custom alternative we make up.

So: we need to decide what it *is* that we're looking for in the cert, and I think we're doing that in the discussion above, but it doesn't need to affect the TCP endpoint address used to reach the kubelet.  I'm worried bringing back Addresses.Hostname will imply that the value is actually a resolvable hostname - and this doesn't exist on all providers.  If we do decide to go with subject == a fake hostname, I'd rather we pulled the fake-hostname from somewhere else (like perhaps nodename).

**Comment by liggitt** [comment]
> We just need something that the CA and master agree they're going to provide and look for in subject/subjectaltname - correct?

Not for verifying the node's serving cert, no. The subject's CN (or a DNS subjectAltName) has to match the hostname used to contact the node, or an IP subjectAltName has to match the IP used to contact the node.

> it doesn't need to affect the TCP endpoint address used to reach the kubelet

Requiring the serving cert to match the TCP endpoint address used to contact the secured server is the point of serving cert validation.

You can make this work if you *always* have a CA that is happy signing serving certs with IP SANs, but that assumption isn't there now, so I'm hesitant to introduce it.

**Comment by javapapo** [comment]
+1

**Comment by luxas** [comment]
+1 for me as well, see #32286

**Comment by bgrant0607** [comment]
https://www.entrust.com/ssl-certificates-without-non-fqdns/

**Comment by anguslees** [comment]
@bgrant0607 that article, and (I believe) @liggitt's objections above assume that the "CA" in question is a public CA.  If I understand the use case correctly here, however, there is no need for the certificate to be verified by a 3rd party, and thus no need for it to be signed by a regular public CA (completely unlike HTTPS in browsers, for example).

The certificates and CA being discussed here are *only* used to secure communication between kubelets and masters, so we control both ends and can make up whatever validation policy we wish.  Again, we just need to decide what it is that we're going to look for in the presented cert, and make sure the CA only signs certificates that follow that policy.

We will *also* want to secure the end-user kubectl/etc clients <-> API servers using regular publicly signed CA certs and follow regular public CA subjectAltName policies, but that bit won't ever involve nodeNames and is not relevant to the discussion in this bug (afaics).

**Comment by liggitt** [comment]
> The certificates and CA being discussed here are only used to secure communication between kubelets and masters

more things talk to the kubelet API than just the apiserver (heapster, for example)

> we control both ends

controlling both the apiserver and the kubelet processes does not mean we control the certificates. the serving certificates are provided by the deployer.

> can make up whatever validation policy we wish.

(copied from https://github.com/kubernetes/kubernetes/pull/25532#issuecomment-250492370): I am strongly opposed to non-standard TLS validation schemes. depending on non-standard cert validation schemes almost guarantees that TLS cert validation will simply be disabled.

**Comment by pesho** [comment]
> Not for verifying the node's serving cert, no. The subject's CN (or a DNS subjectAltName) has to match the hostname used to contact the node, or an IP subjectAltName has to match the IP used to contact the node.

Is it possible to only eliminate NodeName usage for address resolution (by connecting directly to the node IP), but still honor it when checking the TLS cert?

**Comment by liggitt** [comment]
I think the ideal scenario would do something like this:

1. The kubelet records the following addresses, if available:
  * internal IP
  * external IP
  * internal DNS
  * external DNS
2. The apiserver allows prioritizing which address to use when resolving a nodeName to an address

**Comment by liggitt** [comment]
> Is it possible to only eliminate NodeName usage for address resolution (by connecting directly to the node IP), but still honor it when checking the TLS cert?

I don't think we want to continue using nodeName for anything host related. There should be two steps:
* resolve node name to address
* connect to address, expecting a valid TLS cert for that address

**Comment by pesho** [comment]
> * resolve node name to address

I'm confused. This is definitely not what we want. Node name should not be used for any kind of resolution. In many clouds node names are not resolvable.

> * connect to address, expecting a valid TLS cert for that address

This should ideally be "connect to address, expecting a valid TLS cert for that address **or for the NodeName**".

**Comment by liggitt** [comment]
> I'm confused. This is definitely not what we want. Node name should not be used for any kind of resolution. In many clouds node names are not resolvable.

sorry... was unclear. resolve node name to address *by looking up the node object and inspecting the reported addresses in its status*

**Comment by liggitt** [comment]
> This should ideally be "connect to address, expecting a valid TLS cert for that address *or for the NodeName*"

that's the kind of non-standard TLS validation I don't think we should introduce... the serving cert shouldn't be required to contain a CN or SAN unrelated to its serving address.

**Comment by liggitt** [comment]
started a work-in-progress PR at https://github.com/kubernetes/kubernetes/pull/34259 to implement the idea in https://github.com/kubernetes/kubernetes/issues/22063#issuecomment-251997481 (InternalDNS/ExternalDNS address types, letting the apiserver decide which address type to prefer).

I like the `ConnectionInfo` struct in #25532 and the type changes in #33718. I'd be happy for both of those to go in, and I could pick up #34259 afterward. As long as we enable deployers to set up secure apiserver->kubelet TLS over DNS names if they need to, I think we're good.

**Comment by marun** [comment]
For anyone who needs a short-term fix, I'm running ``oc observe`` to synchronize /etc/hosts with nodes' ip address: 

https://github.com/openshift/origin/blob/master/images/dind/master/openshift-sync-etc-hosts.service

**Comment by vjm** [comment]
@marun Anyone have a non-openshift solution?

**Comment by liggitt** [comment]
this can be closed.... fixed in #33718, #35497, and #25532

node.status.addresses are now populated with the hostname the kubelet is given, and the apiserver uses an address reported by the node to contact the kubelet (in the order specified by `--kubelet-preferred-address-types`)

**Comment by vjm** [comment]
@liggitt -- so, if my internal hostname does not resolve correctly I can have communication via IP addresses?

My use case is: in AWS due to use of custom DNS servers, my hostnames for individual machines do not resolve.

**Comment by liggitt** [comment]
yes, start kube-apiserver with `--kubelet-preferred-address-types=InternalIP,ExternalIP`, or whatever priority order you want the apiserver to use to pull addresses out of the node status

**Comment by vjm** [comment]
@liggitt I am not seeing a way to configure this option in `kops`. Do you know of a way to do that?

**Comment by luxas** [comment]
@vjm ask @justinsb or @chrislovecnm or just file a kops issue
